lib/Conversion/StandardToHandshake/StandardToHandshake.cpp

```diff
@@ -1284,68 +1284,66 @@ void connectToMemory(handshake::FuncOp f, MemRefToMemoryAccessOp MemRefOps,
 // A handshake::StartOp should have been created in the first block of the
 // enclosing function region
 static Operation *findStartOp(Region *region) {
-  for (Operation &op : region->front()) {
-    if (isa<handshake::StartOp>(op)) {
+  for (Operation &op : region->front())
+    if (isa<handshake::StartOp>(op))
       return &op;
-    }
-  }
+
   llvm::report_fatal_error("No handshake::StartOp in first block");
 }
 
+template <typename TFuncOp>
+class LowerFuncOpTarget : public ConversionTarget {
+public:
+  explicit LowerFuncOpTarget(MLIRContext &context) : ConversionTarget(context) {
+    loweredFuncs.clear();
+    addLegalDialect<HandshakeOpsDialect>();
+    addLegalDialect<StandardOpsDialect>();
+    /// The root function operation to be replaced is marked dynamically legal
+    /// based on the lowering status of the given function, see
+    /// PartialLowerFuncOp.
+    addDynamicallyLegalOp<TFuncOp>(
+        [&](const auto &funcOp) { return loweredFuncs[funcOp]; });
+  }
```

Comment:
The need to do this is usually because you're using the dialect conversion infrastructure instead of the greedy rewrite infrastructure: https://mlir.llvm.org/docs/PatternRewriter/#common-pattern-drivers  I think this is a case where you want to use applyPatternsAndFold and keep track of whether things are succeeding yourself.